### PR TITLE
[Directize] Add a flag to consider initial table contents immutable

### PR DESCRIPTION
**Pull request by kripken**
In LLVM output and probably others, the initial table contents are never
changed. We may append later, but we don't trample the initial table
entries. As a result, with this new flag we can turn indirect calls on those
offsets into direct ones,
```
--directize-initial-tables-immutable
```

Testing on Emscripten, this definitely helps in practice, e.g. even on
hello world using libc++ it finds 3 opportunities to optimize. However,
those are probably not helpful for speed, and they only save a few
bytes:
```
pop_arg_long_double
fmt_fp
void\20std::__2::__call_once_proxy<std::__2::tuple<std::__2::\28anonymous\20namespace\29::__fake_bind&&>>\28void*\29
```
I'm not sure what the last one is but the first two are internals of
`printf` I believe. Anyhow, in other codebases this may well end up
providing a noticeable improvement.

One limitation though is that we run Directize near the end of the
pipeline, so few followup opts happen after it. We may want to consider
changing that somehow.

### CODE REVIEW

**Comment by MaxGraey** [comment]
Nice! How use this from C-/JS-api? Isn't better add pseudo pass with included this pass arg? Similar to `dae-optimizing`, `simplify-globals-optimizing` and etc

File: src/passes/Directize.cpp [hunk 1]
```diff
@@ -163,11 +206,50 @@ struct Directize : public Pass {
...
+    // This may already be enough information to know that we can't optimize
+    // anything. If so, skip scanning all the module contents.
+    auto canOptimize = [&]() {
+      for (auto& [_, info] : tables) {
+        if (info.canOptimize()) {
+          return true;
+        }
+      }
+      return false;
+    };
```
Comment (sbc100) [inline src/passes/Directize.cpp:245]: Seems like this kind of sub-routine extraction could be done either with a private helper method or with a C++ closure.   Is there some rationale we can/should use for choosing one over the other? 

My understanding is that under the hood the closure does bunch of work, including creating an entire inner class but that we can/should trust the compiler to then eliminate that complexity?   What about debuggability?  Would debugging this in a debugger be any harder than a (arguably more straight forward) private helper method?

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,10 @@ full changeset diff at the end of each section.
 
 Current Trunk
 -------------
+- Add a new flag to Directize, `--directize-initial-tables-immutable` which
```
Comment (sbc100) [inline CHANGELOG.md:17]: When you refer to this argument you don't prefix it with `--pass-arg`.. even though IIUC in practice one always needs that.  How should we document/refer to these flags?

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
> Nice! How use this from C-/JS-api? Isn't better add another pseudo pass with included this pass arg? Similar to `dae-optimizing`, `simplify-globals-optimizing` and etc

Wouldn't we want some way for those API to pass arguments?   There are several passes that take such arguments so it seems to make sense to allow some way to inject them.

**Comment by kripken** [comment]
Yes, if the C API doesn't already have a way to set pass arguments, we should add that.

File: CHANGELOG.md (hunk 2 above)
Comment (kripken) [reply CHANGELOG.md]: Ah, right, this should be prefixed. Fixed in both places in the docs.

File: src/passes/Directize.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/Directize.cpp:245]: I haven't noticed debuggability differences. I think there might be some overhead at least in older compilers here (sometimes they malloc, but not sure when). But overall in a place like this where speed doesn't matter (it's iterating on the tables, which is orders of magnitude smaller than the code) I think it's ok to use a closure, when it is convenient and shorter to write.

**Comment by MaxGraey** [comment]
C-API has BinaryenGetPassArgument, BinaryenSetPassArgument and BinaryenClearPassArguments. But I'm not sure about similar JS api methods

**Pull request merged**